waiter/src/waiter/state.clj

```diff
@@ -1303,3 +1308,114 @@
+                                              (map :id)
+                                              (into healthy-instance-ids))
+                     new-instance-ids (set/difference known-instance-ids' known-instance-ids)
+                     ;; Pair up all of the newly-scheduled instances with their timers.
+                     ;; NOTE: Using split-at allows us to drop the new-instances if we're somehow missing the corresponding timers
+                     [paired-timers scheduling-instance-timer-contexts'] (split-at (count new-instance-ids) scheduling-instance-timer-contexts)
+                     starting-instance-trackers' (->>
+                                                   ;; Start timers for startup time of newly-scheduled instances.
+                                                   ;; NOTE: Since Waiter only scales down by killing known instances, waiting-to-be-scheduled instances are not removed (unlike starting instances).
+                                                   (map (fn [instance-id {schedule-timers :schedule total-timers :total}]
+                                                          (doseq [timer schedule-timers] (timers/stop timer))
+                                                          {:instance-id instance-id
+                                                           :timer-contexts (into [(timers/start (metrics/waiter-timer "service-init-overhead" "startup-time"))
+                                                                                  (timers/start (metrics/service-timer service-id "service-init-overhead" "startup-time"))]
+                                                                                 total-timers)})
+                                                        new-instance-ids paired-timers)
+                                                   (concat starting-instance-trackers)
+                                                   (filterv (fn [{:keys [instance-id timer-contexts]}]
+                                                              (cond
+                                                                ;; Stop tracking any instances that disappeared.
+                                                                (not (contains? known-instance-ids' instance-id)) false
+                                                                ;; Report startup time for any tracked instances that are now healthy.
+                                                                (contains? healthy-instance-ids instance-id) (doseq [tc timer-contexts] (timers/stop tc))
+                                                                ;; Continue tracking all other instances.
+                                                                :else true))))]
+                 (->InstanceTracker service-id known-instance-ids' scheduling-instance-timer-contexts' starting-instance-trackers'))))))
+
+(defn start-instance-startup-stats-maintainer
```

Comment:
1. We can probably move this to `scheduler.clj`.

2. Why not rewrite this to only run on the leader as all routers should effectively report similar metrics? The GC go-blocks are examples of such daemon processes that can preserve state across leadership changes.

Reply:
1. I moved the relevant code loop to `scheduler.clj`; however, now it seems a bit strange to be exposing the loop's internal state via the `/state` endpoint when it's not in `state.clj`.

2. I feel like the extra coordination needed to coordinate only running this on the leader, and persisting the state, is too much complexity for too little benefit. We should let @sradack weigh in on this point too.